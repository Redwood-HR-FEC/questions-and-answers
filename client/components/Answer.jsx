import React from 'react';
import styled from 'styled-components';

const moment = require('moment');

const Date = styled.span`
  color: #767676!important;
`;
const Arrow = styled.li`
  background-image: url(https://m.media-amazon.com/images/G/01/AUIClients/AmazonUIIcon-sprite_1x-e9434610f810a3a99b3f0e49c1b938c9d76628b8._V2_.png);
  background-position-x:-82px;
  background-position-y:-293px;
  background-size:400px 900px;
  display:inline-block;
  color:rgb(17, 17, 17);
  position:relative;
  width:7px;
  height:9px;
  margin-right:0px;
  text-size-adjust:100%;
`;

const ATag = styled.a`
color:rgb(0, 102, 192);
text-decoration-style:solid;
text-decoration-line:none;
line-height:19px;
padding-left: 5px;

&:hover {
  color:rgb(196, 85, 0);
  text-decoration-style:solid;
  text-decoration-line:underline;
}
`;
const AnswerBox = styled.div`
  margin-bottom: 10px;
`;

class Answer extends React.Component {
  constructor({props}) {
    super(props);
    this.state = {
      show: 1,
      open: false,
    };

    this.handleClick = this.handleClick.bind(this);
    this.handleCollapse = this.handleCollapse.bind(this);
  }

  handleClick() {
    // console.log(e);
    const { show, open } = this.state;
    const len = this.props.answer.length;
    let num = show;
    num = show >= len ? show : num + 3;
    // console.log(num);
    this.setState({ show: num, open: !open });
  }

  handleCollapse() {
    let num = 1;
    const { open } = this.state;
    this.setState({ show: num, open: !open });
  }

  render() {
    const { answer } = this.props;
    const { show, open } = this.state;
    let expand = answer.length - show;
    expand = show <= answer.length ? expand : 0;
    let collapse = !open ? '' : <button onClick={this.handleCollapse}>Collapse all answers</button>;
    let clickEvent = (
      <div>
        <Arrow />
        <ATag onClick={this.handleClick}>
          See more answers (
          { expand }
          )
        </ATag>
      </div>
    );

    const showAnswers = answer.slice(0, show).map((a) => (
      <AnswerBox key={a.date_posted + a.username}>
        <span>{a.answer}</span>
        <br />
        <Date>
          By
          {' '}
          {a.username}
          {' '}
          on
          {' '}
          {moment(a.date_posted).format('ll')}
        </Date>
      </AnswerBox>
    ));
    return (
      <div>
        {showAnswers}
        {clickEvent}
        {collapse}
      </div>
    );
  }
}

export default Answer;
