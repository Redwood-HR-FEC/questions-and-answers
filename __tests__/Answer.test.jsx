import React from 'react';
import { shallow } from 'enzyme';
import Answer from '../client/components/Answer';


const sampleAnswers = [
  {
    "username" : "Ahmad.Bruen",
    "answer" : "Ergonomic green workforce Row 24/365 initiative Divide JBOD Money Market Account circuit Ergonomic Wooden Shoes Ergonomic backing up Handcrafted Radial wireless hub optical Up-sized Aruba",
    "date_posted" : "2019-12-07T17:14:54.404Z",
  },
  {
    "username" : "Vivien75",
    "answer" : "Automated deposit silver Strategist Division Berkshire Business-focused Montana deposit Bahamas Rustic Plastic relationships integrated Oklahoma collaborative Chips reboot primary Lebanese Pound",
    "date_posted" : "2019-10-24T15:58:43.321Z",
  },
];

describe('Unit test for Answer component', () => {
  test('Answer component should exists', () => {
    const wrapper = shallow(
      <Answer
        answer={sampleAnswers}
      />,
    );
    expect(wrapper).toExist();
  });

  it("renders correctly", () => {
    const wrapper = shallow(
      <Answer
        answer={sampleAnswers}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Toggle See more', () => {

  test('mouseclick methods', () => {
    const wrapper = shallow(
      <Answer
        answer={sampleAnswers}
      />,
    );
    wrapper.instance().handleClick();
    wrapper.instance().handleCollapse();
  });

  it('should expand to show more items when the See more items is clicked', () => {
    const wrapper = shallow(
      <Answer
        answer={sampleAnswers}
      />,
    );
    // Testing Initial State before a click event
    expect(wrapper.state("open")).toBe(false);
  });

  it ('accepts sampleAnswers props', () => {
    const wrapper = mount(
      <Answer
        answer={sampleAnswers}
      />,
    );
    expect(wrapper.props().answer).toEqual(sampleAnswers);
  });
});
