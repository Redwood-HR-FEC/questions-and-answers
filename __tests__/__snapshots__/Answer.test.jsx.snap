// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for Answer component SnapShot test: renders correctly 1`] = `
<div>
  <Answer__AnswerBox
    key="2019-12-07T17:14:54.404ZAhmad.Bruen"
  >
    <span>
      Ergonomic green workforce Row 24/365 initiative Divide JBOD Money Market Account circuit Ergonomic Wooden Shoes Ergonomic backing up Handcrafted Radial wireless hub optical Up-sized Aruba
    </span>
    <br />
    <Answer__Date>
      By
       
      Ahmad.Bruen
       
      on
       
      Dec 7, 2019
    </Answer__Date>
  </Answer__AnswerBox>
  <div>
    <Answer__Arrow />
    <Answer__ATag
      onClick={[Function]}
    >
      See more answers (
      1
      )
    </Answer__ATag>
  </div>
</div>
`;
